messages/jsonschema/scripts/templates/perl.pm.erb

```diff
@@ -0,0 +1,146 @@
+     line => 12, column => 26
+  );
+  my $loc_json = $loc->to_json;
+
+  my $envelope = Cucumber::Messages::Envelope->from_json($serialized_envelope);
+
+=head1 DESCRIPTION
+
+L<Cucumber messages|https://github.com/cucumber/common/tree/main/messages>
+define the central protocol in the Cucumber ecosystem by which the various
+components communicate. Messages are serialized to NDJSON.
+
+This library provides both serialization/deserialization to/from NDJSON as
+well as the in-memory representation of the messages for Perl applications.
+
+Each serialized message should be wrapped in a C<Cucumber::Messages::Envelope>
+and can thereby be deserialized by calling the C<from_json> class message
+with the serialized representation as its argument, like shown in the SYNOPSIS.
+
+=cut
+
+use strict;
+use warnings;
+
+use Cucumber::Messages::Message;
+
+=head1 MESSAGE CLASSES
+
+=cut
+
+<%- @schemas.sort.each do |key, schema| %>
+
+package Cucumber::Messages::<%= class_name(key) %> {
+
+=head2 Cucumber::Messages::<%= class_name(key) %>
+
+=head3 DESCRIPTION
+
+Represents the <%= class_name(key) %> message in Cucumber's
+L<message protocol|https://github.com/cucumber/common/tree/main/messages#readme>.
+
+<%= format_description(schema['description']) %>
+
+=head3 ATTRIBUTES
+
+=cut
+
+use Moo;
+extends 'Cucumber::Messages::Message';
+
+use Scalar::Util qw( blessed );
+
+my %types = (
+<%- schema['properties'].each do |property_name, property| -%>
+   <%= underscore(property_name) %> => '<%= type_for(class_name(key), property_name, property) %>',
+<%- end -%>
+);
+
+# This is a work-around for the fact that Moo doesn't have introspection
+# and Perl doesn't have boolean values...
+sub _types {
+    return \%types;
+}
+
+
+<%- schema['properties'].each do |property_name, property| -%>
+
+=head4 <%= underscore(property_name) %>
+
+<%- if property['description'] -%>
+<%= format_description(property['description']) %>
+<%- end -%>
+
+<% if property['enum'] -%>
+
+Available constants for valid values of this field:
+
+=over
+
+<%- property['enum'].to_a.each_with_index do |value, index| -%>
+=item * <%= property_name.upcase %>_<%= enum_constant(value) %>
+
+<%- end -%>
+=back
+
+<% end -%>
+=cut
+
+<% if property['enum'] -%>
+
+use constant
+<%- property['enum'].to_a.each_with_index do |value, index| -%>
+   <%= property_name.upcase %>_<%= enum_constant(value) %> => '<%= value %>',
+<%- end -%>
+   ;
+
+<% end -%>
+has <%= underscore(property_name) %> =>
+    (is => 'ro',
+     <%- if (schema['required'] || []).index(property_name) -%>
+     required => 1,
+     <%- if default_value(class_name(key), property_name, property) -%>
+     default => sub { <%= default_value(class_name(key), property_name, property) %> },
+     <%- end -%>
+     <%- end -%>
+    );
+
+<%- end %>
+}<%- end -%>
+
+
+1;
+
+__END__
+
+=head1 LICENSE
+
+Please see the included LICENSE.txt for the canonical version. In summary:
+
+The MIT License (MIT)
+
+Copyright (c) 2021 Erik Huelsmann
```

Comment:
I think we might need to change this to something else that represents to Cucumber Open community.

We’ve been using Cucumber Ltd for other repos, but that company is defunkt. I’m not sure we ahould use SmartBear either, since it might deter people from contributing.

@mattwynne @aurelien-reeves what do you think? Time to revisit the license owner of all the cucumber/* repos?

Reply:
Let me know what it should be. if there were a legal body, we would be able to work with ICLA or CCLA constructs like the Apache Foundation does. That would be the  most sound solution. I can change it to "The Cucumber Project", but maybe that's just an empty shell/meaningless term at the moment?

Reply:
@mattwynne @aslakhellesoy maybe it's consistent for now to (at least) add Cucumber Ltd for now? That way, it's consistent with the other repositories.

Reply:
That would be great - thanks. Then we can address the license owner issue on a broader scale later.

Reply:
Ok. I actually already set the copyright holder in the metadata to Cucumber Ltd: https://github.com/cucumber/common/pull/1670/files#diff-e4960b209d869c770dab6dae20e4e07a5823e193b23d01967012993cde5c907eR12 I'll submit an additional commit to add Cucumber Ltd to the author/copyright in the generated messages file.

Reply:
How about this instead:

```perl
=head1 LICENSE

Please see the included LICENSE for the canonical version. In summary:

The MIT License (MIT)

  Copyright (c) 2021 Erik Huelsmann
  Copyright (c) 2021 Cucumber Ltd

This work is loosely derived from prior work of the same library for Ruby,
called C<cucumber-messages>.

=cut
```

?

Reply:
I guess my question is: this seems to be in line with the Gherkin/Perl module; maybe that was added in another era though?

Reply:
Cucumber Ltd is wholly owned by SmartBear, so really they're interchangeable. I can't comment on whether it would deter people or not.

I don't really understand what it means to copyright something with an MIT license anyway. Does it enable you to sue someone if they tried to claim they made it instead of you?

Reply:
As the copyright holder, you're allowed to change the license as you own the creative rights to the code. With the MIT license, the distributor and user can do with it what they want (even create derived works), but they can't change the license without the consent of the original copyright holders.

Reply:
I think this is good enough for now.